loopy/check.py

```diff
@@ -1306,16 +1306,21 @@ def check_that_temporaries_are_defined_in_subkernels_where_used(kernel):
     from loopy.kernel.data import AddressSpace
     from loopy.kernel.tools import get_subkernels
 
+    globally_defined_base_storage = set()
+
     for subkernel in get_subkernels(kernel):
-        defined_base_storage = set()
+        locally_defined_base_storage = set()
 
         from loopy.schedule.tools import (
                 temporaries_written_in_subkernel, temporaries_read_in_subkernel)
 
         for temporary in temporaries_written_in_subkernel(kernel, subkernel):
             tval = kernel.temporary_variables[temporary]
-            if tval.base_storage is not None:
-                defined_base_storage.add(tval.base_storage)
+            if tval.address_space in (AddressSpace.PRIVATE, AddressSpace.LOCAL):
+                locally_defined_base_storage.add(tval.base_storage)
+            else:
+                assert tval.address_space == AddressSpace.GLOBAL
+                globally_defined_base_storage.add(tval.base_storage)
```

Comment:
Won't this spuriously add `None` to `*_defined_base_storage`?

Reply:
Yep thanks! Luckily the usage site would always check for a 'non-None' base_storage. Fixed it by adding the following diff to the patch:
```diff
diff --git a/loopy/check.py b/loopy/check.py
index 3d7d986bc..9634ba6a0 100644
--- a/loopy/check.py
+++ b/loopy/check.py
@@ -1316,11 +1316,12 @@ def check_that_temporaries_are_defined_in_subkernels_where_used(kernel):
 
         for temporary in temporaries_written_in_subkernel(kernel, subkernel):
             tval = kernel.temporary_variables[temporary]
-            if tval.address_space in (AddressSpace.PRIVATE, AddressSpace.LOCAL):
-                locally_defined_base_storage.add(tval.base_storage)
-            else:
-                assert tval.address_space == AddressSpace.GLOBAL
-                globally_defined_base_storage.add(tval.base_storage)
+            if tval.base_storage is not None:
+                if tval.address_space in (AddressSpace.PRIVATE, AddressSpace.LOCAL):
+                    locally_defined_base_storage.add(tval.base_storage)
+                else:
+                    assert tval.address_space == AddressSpace.GLOBAL
+                    globally_defined_base_storage.add(tval.base_storage)
```